metagenomics.py

```diff
@@ -716,62 +734,68 @@ def parser_krona(parser=argparse.ArgumentParser()):
+        '/dev/stdout',
+        interleave=True,
+        illuminaClipping=True,
+        JVMmemory=picard.jvmMemDefault,
+        background=True,
+        stdout=subprocess.PIPE,
     )
 
     diamond_tool = tools.diamond.Diamond()
     diamond_tool.install()
-    tmp_alignment = util.file.mkstempfname('.daa')
-    tmp_m8 = util.file.mkstempfname('.diamond.m8')
-    diamond_tool.blastx(db, [tmp_fastq, tmp_fastq2], tmp_alignment, options={'--threads': numThreads})
-    diamond_tool.view(tmp_alignment, tmp_m8, options={'--threads': numThreads})
-
-    if outM8:
-        with open(tmp_m8, 'rb') as f_in:
-            with gzip.open(outM8, 'wb') as f_out:
-                shutil.copyfileobj(f_in, f_out)
+    taxonmap = join(taxDb, 'accession2taxid', 'prot.accession2taxid.gz')
+    taxonnodes = join(taxDb, 'nodes.dmp')
+    cmd = '{} blastx --db {db}'.format(diamond_tool.install_and_get_path(), db=db)
 
-    tax_db = TaxonomyDb(tax_dir=taxDb, load_names=True, load_nodes=True, load_gis=True)
-    tmp_lca_tsv = util.file.mkstempfname('.tsv')
-    with open(tmp_m8) as m8, open(tmp_lca_tsv, 'w') as lca:
-        blast_lca(tax_db, m8, lca, paired=True, min_bit_score=50)
+    cmd = '{} blastx --outfmt 102 --sallseqid'.format(diamond_tool.install_and_get_path())
+    cmd += ' --threads {threads}'.format(threads=numThreads)
+    cmd += ' --db {db} --taxonmap {taxonmap} --taxonnodes {taxonnodes}'.format(
+        threads=numThreads,
+        db=db,
+        taxonmap=taxonmap,
+        taxonnodes=taxonnodes)
 
-    if outLca:
-        with open(tmp_lca_tsv, 'rb') as f_in:
-            with gzip.open(outLca, 'wb') as f_out:
-                shutil.copyfileobj(f_in, f_out)
+    if outLca is not None:
+        cmd += ' | tee >(gzip > {out})'.format(out=outLca)
+
+    diamond_ps = subprocess.Popen(cmd, shell=True, stdin=subprocess.PIPE,
+                                  stdout=subprocess.PIPE, executable='/bin/bash')
+
+    def f(input_pipe, output_pipe):
+        output_pipe = codecs.getwriter('ascii')(output_pipe)
+        SeqIO.write(util.file.join_interleaved_fastq(input_pipe, output_format='fasta', num_n=16),
+                    output_pipe, 'fasta')
+
+    util.misc.bind_pipes(s2fq.stdout, diamond_ps.stdin, f)
 
-    with open(tmp_lca_tsv) as f:
-        hits = taxa_hits_from_tsv(f)
+    tax_db = TaxonomyDb(tax_dir=taxDb, load_names=True, load_nodes=True)
+
+    lca_p = codecs.getreader('ascii')(diamond_ps.stdout)
+
+    hits = taxa_hits_from_tsv(lca_p)
     with open(outReport, 'w') as f:
         for line in kraken_dfs_report(tax_db, hits):
             print(line, file=f)
 
+    s2fq.wait()
+    diamond_ps.wait()
+
 
 def parser_diamond(parser=argparse.ArgumentParser()):
     parser.add_argument('inBam', help='Input unaligned reads, BAM format.')
     parser.add_argument('db', help='Diamond database directory.')
     parser.add_argument('taxDb', help='Taxonomy database directory.')
     parser.add_argument('outReport', help='Output taxonomy report.')
```

Comment:
Perhaps for consistency with `parser_kraken`, should we rename `--outLca` to `--outReads`, and choose something consistent for both kraken and diamond in terms of which of required vs optional on the two output files?

Reply:
Done